R/pmcmc.R

```diff
@@ -0,0 +1,609 @@
+##'
+##' @param run_params List of parameters for \code{particle_filter$run}
+##'
+##' @param output_proposals Logical indicating whether proposed parameter
+##' jumps should be output along with results
+##'
+##' @param n_chains Number of chains to run
+##'
+##' @return an mcmc object containing
+##' - List of inputs
+##' - Matrix of accepted parameter samples, rows = iterations
+##'   as well as log prior, (particle filter estimate of) log likelihood
+##'   and log posterior
+##'
+##' @description This is a basic Metropolis-Hastings MCMC sampler.
+##' The \code{filter} is run with a set of parameters to evaluate the
+##' likeihood. A new set of parameters is proposed, and these likelihoods are
+##' compared, jumping with probability equal to their ratio. This is repeated
+##' for \code{n_mcmc} proposals, \code{n_chains} independent times.
+##'
+##' @export
+##' @import coda
+##' @import furrr
+##' @importFrom stats rnorm
+##' @importFrom mvtnorm rmvnorm
+##' @importFrom graphics matplot
+pmcmc <- function(mcmc_range,
+                  lprior_funcs,
+                  filter,
+                  n_particles,
+                  n_mcmc,
+                  proposal_kernel,
+                  run_params = NULL,
+                  output_proposals = FALSE,
+                  n_chains = 1) {
+  vars <- mcmc_validate_range(mcmc_range)
+  par_names <- as.character(vars$range$name)
+
+  if (length(output_proposals) != 1 || !is.logical(output_proposals)) {
+    stop("output_proposals must be either TRUE or FALSE")
+  }
+
+  if (!all(par_names %in% names(lprior_funcs))) {
+    stop("All sampled parameters must have a defined prior")
+  }
+
+  if (!(setequal(rownames(proposal_kernel),
+                colnames(proposal_kernel)) &&
+         setequal(rownames(proposal_kernel),
+                  par_names) &&
+      nrow(proposal_kernel) == length(par_names) &&
+      ncol(proposal_kernel) == length(par_names))) {
+    stop(paste0("proposal_kernel must be a matrix or vector with names ",
+                "corresponding to the parameters being sampled"))
+  }
+
+  # For the chains to be independent on threads, dust will
+  # need to support a long_jump call
+  # (for now, could use seed = seed + thread_idx)
+  chains <- furrr::future_pmap(
+      .l =  list(n_mcmc = rep(n_mcmc, n_chains)),
+      .f = run_mcmc_chain,
+      vars,
+      lprior_funcs,
+      filter,
+      n_particles,
+      proposal_kernel,
+      run_params = run_params,
+      output_proposals = output_proposals,
+      .progress = TRUE)
+
+  if (n_chains > 1) {
+    names(chains) <- paste0("chain", seq_len(n_chains))
+
+    # calculating rhat
+    # convert parallel chains to a coda-friendly format
+    chains_coda <- lapply(chains, function(x) {
+        traces <- x$results
+      coda::as.mcmc(traces[, vars$range$name])
+    })
+
+    rhat <- tryCatch(expr = {
+      x <- coda::gelman.diag(chains_coda)
+      x
+    }, error = function(e) {
+      print("unable to calculate rhat")
```

Comment:
print is the worst named function in R - please use message. Also this error handler should probably return `NA_real_` or `NULL` as it is currently returning the string!